neo4j/driver_test.go

```diff
@@ -84,158 +80,210 @@ func assertStringContain(t *testing.T, str, substr string) {
-			driverTarget := driver.Target()
+	t.Run("neo4j:// routing context with duplicate keys should error", func(t1 *testing.T) {
+		_, err := NewDriver("neo4j://localhost:7687?x=y&x=b", NoAuth())
 
-			assertNoError(t, err)
-			assertStringEqual(t, driverTarget.Port(), "7687")
-		})
+		assertError(t, err)
 	})
-}
 
-var _ = Describe("Driver", func() {
-	When("constructed with bolt://localhost:7687", func() {
-		driver, err := NewDriver("bolt://localhost:7687", NoAuth())
+}
 
-		It("no errors should be returned", func() {
-			Expect(err).To(BeNil())
-		})
-
-		When("closed", func() {
-			err := driver.Close()
-			It("no errors should be returned", func() {
-				Expect(err).To(BeNil())
-			})
-
-			It("should not allow new read mode sessions", func() {
-				session, err := driver.Session(AccessModeRead)
-				Expect(err).NotTo(BeNil())
-				Expect(session).To(BeNil())
-			})
-
-			It("should not allow new write mode sessions", func() {
-				session, err := driver.Session(AccessModeWrite)
-				Expect(err).NotTo(BeNil())
-				Expect(session).To(BeNil())
-			})
-
-			It("should allow more close calls", func() {
-				err := driver.Close()
-				Expect(err).To(BeNil())
-			})
-		})
+func TestDriverDefaultPort(t *testing.T) {
 
+	t.Run("neo4j://localhost should default to port 7687", func(t1 *testing.T) {
+		driver, err := NewDriver("neo4j://localhost?x=y&a=b", NoAuth())
 		driverTarget := driver.Target()
 
-		Context("driver's", func() {
-			It("target scheme should be bolt", func() {
-				Expect(driverTarget.Scheme).To(BeIdenticalTo("bolt"))
-			})
+		assertNoError(t1, err)
+		assertStringEqual(t1, driverTarget.Port(), "7687")
+	})
+
+}
+
+func TestNewDriverAndClose(t *testing.T) {
+
+	driver, err := NewDriver("bolt://localhost:7687", NoAuth())
+	assertNoError(t, err)
+
+	driverTarget := driver.Target()
+
+	if driverTarget.Scheme != "bolt" {
+		t.Errorf("the URI scheme was not properly set %v", driverTarget.Scheme)
+	}
+
+	if driverTarget.Hostname() != "localhost" {
+		t.Errorf("the hostname is not the expected %v", driverTarget.Hostname())
+	}
+
+	if driverTarget.Port() != "7687" {
+		t.Errorf("the port is not the expected %v", driverTarget.Port())
+	}
+
+	err = driver.Close()
+	assertNoError(t, err)
+
+	session, err := driver.Session(AccessModeRead)
+	assertError(t, err)
+
+	if session != nil {
+		t.Errorf("should not allow new session (AccessModeRead) after driver being closed")
+	}
+
+	session, err = driver.Session(AccessModeWrite)
+	assertError(t, err)
 
-			It("target hostname should be localhost", func() {
-				Expect(driverTarget.Hostname()).To(BeIdenticalTo("localhost"))
-			})
+	if session != nil {
+		t.Errorf("should not allow new session (AccessModeWrite) after driver being closed")
+	}
+
+	err = driver.Close()
+	if err != nil {
+		t.Errorf("should allow the close call on a closed driver")
+	}
+
+}
+
+func TestDriverSessionCreation(t *testing.T) {
+
+	t.Run("case one", func(t1 *testing.T) {
```

Comment:
These looks like they should be table driven. Look here for an example: https://github.com/golang/go/wiki/TableDrivenTests
There are also examples within this codebase (probably within internal/...).
The auth stuff could also easily be written with thie approach, just comparing maps instead of comparing each entry in the maps

Reply:
updated